cpp/include/cudf/detail/join/filtered_join.cuh

```diff
@@ -0,0 +1,359 @@
+#pragma once
+
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/join/join.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/primitive_row_operators.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cuco/bucket_storage.cuh>
+#include <cuco/extent.cuh>
+#include <cuco/static_multiset_ref.cuh>
+#include <cuco/static_set_ref.cuh>
+#include <cuco/types.cuh>
+#include <cuda/std/type_traits>
+
+#include <type_traits>
+
+// Forward declaration
+namespace cudf::experimental::row::equality {
+class preprocessed_table;
+}
+
+namespace cudf {
+namespace detail {
+
+using cudf::experimental::row::lhs_index_type;
+using cudf::experimental::row::rhs_index_type;
+
+/**
+ * @brief Base class providing common functionality for filtered join operations.
+ *
+ * This abstract class implements the core components needed for hash-based semi
+ * and anti join operations.
+ */
+class filtered_join {
+ public:
+  /**
+   * @brief Properties of the build table used in the join operation
+   */
+  struct build_properties {
+    bool _has_nulls;           ///< True if nested nulls are present in build table
+    bool _has_floating_point;  ///< True if the build table contains floating point columns
+    bool _has_nested_columns;  ///< True if the build table contains nested columns
+  };
+
+  /**
+   * @brief Adapter for insertion operations in the hash table
+   *
+   * When build table is reused, always returns false to allow duplicate entries in the hash table
+   * for no-op comparison. When build table is not reused, returns result of self comparator passed
+   */
+  template <typename T>
+  struct insertion_adapter {
+    insertion_adapter(T const& _c) : _comparator{_c} {}
+    __device__ constexpr bool operator()(
+      cuco::pair<hash_value_type, lhs_index_type> const& lhs,
+      cuco::pair<hash_value_type, lhs_index_type> const& rhs) const noexcept
+    {
+      if constexpr (std::is_same_v<T, bool>) {
+        return false;
+      } else {
+        if (lhs.first != rhs.first) { return false; }
+        auto const lhs_index = static_cast<size_type>(lhs.second);
+        auto const rhs_index = static_cast<size_type>(rhs.second);
+        return _comparator(lhs_index, rhs_index);
+      }
+    }
+
+   private:
+    T _comparator;
+  };
+
+  /**
+   * @brief Adapter for extracting hash values from key-value pairs
+   */
+  struct hasher_adapter {
+    template <typename T>
+    __device__ constexpr hash_value_type operator()(
+      cuco::pair<hash_value_type, T> const& key) const noexcept
+    {
+      return key.first;
+    }
+  };
+
+  /**
+   * @brief Adapter for generating key-value pairs from indices
+   *
+   * @tparam T Index type
+   * @tparam Hasher Hash function type
+   */
+  template <typename T, typename Hasher>
+  struct keys_adapter {
```

Comment:
The name `keys_adapter` sounds confusing to me. If this is to generate a `key` pair, calling it `key_pair_fn` should be better.